### PR TITLE
ENG-11110: Add a dummy transaction task message at the end of SP lead…

### PR DESCRIPTION
**Pull request by xinjiacs**
…er promotion. The request will be duplicated to replicas like a SP write. It does not do anything useful work, but as a signal to flush everyone's transaction queue before the newly promoted leader starting to run new transactions.

### CODE REVIEW

File: src/frontend/org/voltdb/iv2/DummyTransactionTask.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.io.IOException;
+
+import org.voltcore.messaging.Mailbox;
+import org.voltdb.SiteProcedureConnection;
+import org.voltdb.dtxn.TransactionState;
+import org.voltdb.messaging.DummyTransactionResponseMessage;
+import org.voltdb.rejoin.TaskLog;
+
+public class DummyTransactionTask extends TransactionTask {
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/DummyTransactionTask.java]: Should add a comment here explaining what this is for.

File: src/frontend/org/voltdb/iv2/InitiatorMailbox.java [hunk 2]
```diff
@@ -38,9 +38,9 @@
 import org.voltdb.messaging.Iv2RepairLogRequestMessage;
 import org.voltdb.messaging.Iv2RepairLogResponseMessage;
 import org.voltdb.messaging.RejoinMessage;
+import org.voltdb.messaging.RepairLogTruncationMessage;
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/InitiatorMailbox.java]: This change is completely unnecessary.

File: src/frontend/org/voltdb/iv2/SpInitiator.java [hunk 3]
```diff
@@ -175,6 +176,12 @@ public void acceptPromotion()
                     LeaderCacheWriter iv2masters = new LeaderCache(m_messenger.getZK(),
                             m_zkMailboxNode);
                     iv2masters.put(m_partitionId, m_initiatorMailbox.getHSId());
+
+                    // After SP leader promotion, a DummyTransactionTaskMessage is generated from the new leader.
+                    // This message will be put in transaction task queue and duplicated for response from every replica.
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpInitiator.java]: "duplicated for response from every replica" is confusing.

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 4]
```diff
@@ -411,7 +419,7 @@ private long getMaxTaskedSpHandle() {
 
     // SpScheduler expects to see InitiateTaskMessages corresponding to single-partition
     // procedures only.
-    public void handleIv2InitiateTaskMessage(Iv2InitiateTaskMessage message)
+    private void handleIv2InitiateTaskMessage(Iv2InitiateTaskMessage message)
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: These changes are completely unnecessary.

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 5]
```diff
@@ -1143,6 +1151,61 @@ public void handleDumpMessage()
...
     }
 
+    private void handleDummyTransactionTaskMessage(DummyTransactionTaskMessage message)
+    {
+        DummyTransactionTaskMessage msg = message;
+        long uniqueId, newSpHandle;
+        if (m_isLeader) {
+            TxnEgo ego = advanceTxnEgo();
+            newSpHandle = ego.getTxnId();
+            uniqueId = m_uniqueIdGenerator.getNextUniqueId();
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: You don't really need a unique ID for this message.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+public class DummyTransactionTaskMessage extends TransactionInfoBaseMessage
+{
+    public DummyTransactionTaskMessage()
+    {
+        super();
+        m_isReadOnly = true;
+        m_isForReplay = false;
+    }
+
+    public DummyTransactionTaskMessage (long initiatorHSId, long coordinatorHSId, long txnId, long uniqueId) {
```
Comment (nshi) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: Isn't `initiatorHSId` and `coordinatorHSId` always the same? If so, one of the parameters can go away.

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 7]
```diff
@@ -1143,6 +1151,61 @@ public void handleDumpMessage()
...
+
+            msg = new DummyTransactionTaskMessage(m_mailbox.getHSId(),
+                    m_mailbox.getHSId(), newSpHandle, uniqueId);
+            msg.setSpHandle(newSpHandle);
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: Why set the spHandle right after you passed it in to the constructor?

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 8]
```diff
@@ -1143,6 +1151,61 @@ public void handleDumpMessage()
...
+                safeAddToDuplicateCounterMap(new DuplicateCounterKey(msg.getTxnId(), newSpHandle), counter);
+            }
+        } else {
+            setMaxSeenTxnId(msg.getSpHandle());
+        }
+        Iv2Trace.logDummyTransactionTaskMessage(msg, m_mailbox.getHSId());
+
+        DummyTransactionTask task = new DummyTransactionTask(m_mailbox,
+                new SpTransactionState(msg), m_pendingTasks);
+        m_pendingTasks.offer(task);
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: You need to pass it to the command logging thread to flush the queue there, too. It can't go into the transaction task queue directly.

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 9]
```diff
@@ -1143,6 +1151,61 @@ public void handleDumpMessage()
...
+    }
+
+    private void handleDummyTransactionResponseMessage(DummyTransactionResponseMessage message) {
+        final long spHandle = message.getSpHandle();
+        final DuplicateCounterKey dcKey = new DuplicateCounterKey(message.getTxnId(), spHandle);
+        DuplicateCounter counter = m_duplicateCounters.get(dcKey);
+        if (counter == null) {
+            // this will be on SPI without k-safety or replica only with k-safety
+            setRepairLogTruncationHandle(spHandle);
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: Since the dummy transaction is a read-only one, you shouldn't advance the spHandle for the dummy message. That saves you the trouble of advancing the truncation handle.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+import org.voltcore.messaging.TransactionInfoBaseMessage;
+import org.voltcore.utils.CoreUtils;
+import org.voltdb.iv2.TxnEgo;
+import org.voltdb.iv2.UniqueIdGenerator;
+
+/**
+ * Message from a client interface to an initiator, instructing the
+ * site to begin executing a stored procedure, coordinating other
+ * execution sites if needed.
```
Comment (nshi) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: Please fix copy/pasted comment.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        m_isForReplay = false;
+    }
+
+    @Override
+    public boolean isSinglePartition() {
+        return true;
+    }
+
+    @Override
+    public int getSerializedSize()
```
Comment (nshi) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: You don't have to override it if you are not changing the implementation.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        buf.put(VoltDbMessageFactory.DUMMY_TRANSACTION_TASK_ID);
+        super.flattenToBuffer(buf);
+
+        assert(buf.capacity() == buf.position());
+        buf.limit(buf.position());
+    }
+
+    @Override
+    public void initFromBuffer(ByteBuffer buf) throws IOException {
```
Comment (nshi) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: Same here.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 9 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: Not really here for this case. The problem we are fixing is the next new SP RO txn will be released immediately because of the early committed last repaired SP. When the dummy txn  advances the SpHandle, the following new SP RO txns will not be released until the dummy txn gets committed (when SPI gets acked from all replicas by dummy responses).

File: src/frontend/org/voltdb/iv2/DummyTransactionTask.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    final Mailbox m_initiator;
+
+    public DummyTransactionTask(Mailbox initiator,
+            TransactionState txnState, TransactionTaskQueue queue) {
+        super(txnState, queue);
+        m_initiator = initiator;
+    }
+
+    private void generateDummyResponse() {
+        DummyTransactionResponseMessage response = new DummyTransactionResponseMessage(this);
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/DummyTransactionTask.java]: You also have to mark the transaction state as done. Otherwise, `SpScheduler.updateReplicas()` won't repair it correctly, which may deadlock the system.

File: src/frontend/org/voltdb/iv2/SpInitiator.java (hunk 3 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/SpInitiator.java]: I updated this message. Check it again. 
I still have not proved that we have a bug not picking the longest command log to replay.

File: src/frontend/org/voltdb/iv2/DummyTransactionTask.java (hunk 13 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/DummyTransactionTask.java]: Good catch.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 4 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: I think it's better being private.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 7 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: The reason is the super constructor sets SpHandle member variable as Long.MIN_VALUE as default. 

If uniqueId is not used, we can put this setter function inside of the constructor of DummyTransactionTaskMessage.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 8 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: The DummyTransactionTask is a read only task, why it needs to be command logged?

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java (hunk 6 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: Right. I guess it's the legacy of copy/past.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java (hunk 11 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: Initially when I added extra member variables, I need it. After removing them, I end up leaving it there. Good catch.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 8 above)
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: It doesn't need to be command logged, but you have to flush the command log queue. Otherwise, there may be transactions sitting in the command log queue and you'll miss them and run into the same bug again.

File: src/frontend/org/voltdb/iv2/SpScheduler.java (hunk 7 above)
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: You can set the spHandle directly in the constructor. Unique ID is not needed for this message.

File: src/frontend/org/voltdb/iv2/InitiatorMailbox.java (hunk 2 above)
Comment (nshi) [inline src/frontend/org/voltdb/iv2/InitiatorMailbox.java]: Yes, it's better to organize them together. On the other hand, this is completely unnecessary for this bug fix. I recommend to **not** reorganize imports because it generates a lot of unnecessary changes in commits. For example, if my IDE thinks it's better to organize imports in a slightly different way than yours, we'll constantly generate unnecessary commit changes when we touch the same files.

File: src/frontend/org/voltdb/iv2/SpScheduler.java [hunk 14]
```diff
@@ -1154,15 +1154,12 @@ private void handleDumpMessage()
     private void handleDummyTransactionTaskMessage(DummyTransactionTaskMessage message)
     {
         DummyTransactionTaskMessage msg = message;
-        long uniqueId, newSpHandle;
+        long newSpHandle;
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/SpScheduler.java]: This variable can go inside the leader if branch.

File: src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java [hunk 15]
```diff
@@ -23,13 +23,10 @@
...
 import org.voltdb.iv2.TxnEgo;
-import org.voltdb.iv2.UniqueIdGenerator;
 
 /**
- * Message from a client interface to an initiator, instructing the
- * site to begin executing a stored procedure, coordinating other
- * execution sites if needed.
- *
+ * Message from an initiator just after newly leader promotion, instructing the
+ * site to begin executing a stored procedure, coordinating other execution sites if needed.
```
Comment (nshi) [inline src/frontend/org/voltdb/messaging/DummyTransactionTaskMessage.java]: This comment is still incorrect. Please read the whole sentence.
